File: typescript/src/data-structures/trie.ts
import { NaryTree, NaryTreeNode, isDummy } from "./tree";

export class Trie extends NaryTree<string> {
  constructor() {
    super();
    this.root = {
      value: "",
      parent: undefined,
      children: []
    };
  }

  insert(word: string) {
    if (/\s/.test(word)) {
      throw new Error(`Invalid word: ${word}`);
    }
    const letterArray = [...word];
    let currentNode = this.root;
    let didInsert = false;
    letterArray.forEach(letter => {
      const letterPosition = currentNode.children.findIndex(child => !isDummy(child) && child.value === letter);

      if (letterPosition > -1) {
        currentNode = currentNode.children[letterPosition] as NaryTreeNode<string>;
      } else {
        currentNode.children.push({ value: letter, parent: currentNode, children: [] });
        currentNode = currentNode.children[currentNode.children.length - 1] as NaryTreeNode<string>;
        this._size++;
        didInsert = true;
      }
    });
    if (didInsert || !currentNode.children.find(node => isDummy(node))) {
      currentNode.children.push({ nodeType: "dummy" });
      return true;
    }
    return false;
  }

  /** Checks if a prefix exists in the trie. */
  validatePrefix(prefix: string) {
    const letterArray = [...prefix];
    let currentNode = this.root;
    for (const letter of letterArray) {
      const letterPosition = currentNode.children.findIndex(child => !isDummy(child) && child.value === letter);
      if (letterPosition === -1) {
        return false;
      }
      currentNode = currentNode.children[letterPosition] as NaryTreeNode<string>;
    }
    return true;
  }

  /** Checks if a word exists in the trie. */
  validateWord(word: string) {
    const letterArray = [...word];
    let currentNode = this.root;
    for (const letter of letterArray) {
      const letterPosition = currentNode.children.findIndex(child => !isDummy(child) && child.value === letter);
      if (letterPosition === -1) {
        return false;
      }
      currentNode = currentNode.children[letterPosition] as NaryTreeNode<string>;
    }
    return !!currentNode.children.find(node => isDummy(node));
  }

  preOrder() {
    // Remove the first node in the traversal since the root is empty.
    return super.preOrder().slice(1);
  }

  postOrder() {
    // Remove the last node in the traversal since the root is empty.
    return super.postOrder().slice(0, this.size);
  }
}

File: ruby/src/data-structures/trie.rb
# frozen-string-literal: true

require 'tree'

module DataStructures
  class Trie
    attr_reader :size

    def initialize
      @root = NaryTreeNode.new('', nil, [])
      @size = 0
    end

    def insert(word)
      return raise "Invalid word: #{word}" if word =~ /\s/

      letters = word.split('')
      current = @root
      did_insert = false
      letters.each do |letter|
        letter_position = current.children.index(current.children.find { |x| !dummy?(x) && x.value == letter })

        if !letter_position.nil?
          current = current.children[letter_position]
        else
          current.children.push(NaryTreeNode.new(letter, current, []))
          current = current.children.last
          @size += 1
          did_insert = true
        end
      end

      return false unless did_insert || current.children.find { |node| dummy?(node) }.nil?

      dummy = NaryTreeNode.new('dummy', current, [])
      dummy.node_type = 'dummy'
      current.children.push(dummy)
      true
    end

    def insert_all(elems)
      result = true
      elems.each do |elem|
        success = insert(elem)
        result &&= success
      end
      result
    end

    def validate_prefix(prefix)
      letters = prefix.split('')
      current = @root
      letters.each do |letter|
        letter_position = current.children.index(current.children.find { |x| !dummy?(x) && x.value == letter })
        return false if letter_position.nil?

        current = current.children[letter_position]
      end
      true
    end

    def validate_word(word)
      letters = word.split('')
      current = @root
      letters.each do |letter|
        letter_position = current.children.index(current.children.find { |x| !dummy?(x) && x.value == letter })
        return false if letter_position.nil?

        current = current.children[letter_position]
      end
      !current.children.find { |node| dummy?(node) }.nil?
    end

    def clear
      @root = NaryTreeNode.new('', nil, [])
      @size = 0
    end

    def empty?
      @size.zero?
    end

    private

    def dummy?(node)
      node.node_type == 'dummy'
    end
  end
end
